src/hotspot/os/bsd/memMapPrinter_macosx.cpp

```diff
@@ -0,0 +1,383 @@
+    }
+
+    st->print_raw(mapping_info._file_name.base());
+    st->cr();
+
+#undef INDENT_BY
+  }
+
+  void print_legend() const {
+    outputStream* st = _session.out();
+    st->print_cr("from, to, vsize: address range and size");
+    st->print_cr("prot:    protection:");
+    st->print_cr("           rwx: read / write / execute");
+    st->print_cr("share:   share mode:");
+    st->print_cr("           cow: copy on write");
+    st->print_cr("           pvt: private");
+    st->print_cr("           shr: shared");
+    st->print_cr("           tsh: true shared");
+    st->print_cr("           p/a: private aliased");
+    st->print_cr("           s/a: shared aliased");
+    st->print_cr("           lpg: large page");
+    st->print_cr("offset:  offset from start of allocation block");
+    st->print_cr("vminfo:  VM information (requires NMT)");
+    st->print_cr("file:    file mapped, if mapping is not anonymous");
+    {
+      streamIndentor si(st, 16);
+      _session.print_nmt_flag_legend();
+    }
+    st->print_cr("file:            file mapped, if mapping is not anonymous");
+  }
+
+  void print_header() const {
+    outputStream* st = _session.out();
+    //            0         1         2         3         4         5         6         7         8         9         0         1         2         3         4         5         6         7
+    //            012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789
+    //            0x000102890000-0x000102898000                32768 r--/r-- cow       0xc000 /Users/simont/dev/openjdk/jdk/build/macos-aarch64-fastdebug-shenandoah/images/jdk/bin/java
+    st->print_cr("from               to                        vsize prot    share     offset  vminfo/file");
+    st->print_cr("==================================================================================================");
+  }
+};
+
+static bool is_interesting(const proc_regionwithpathinfo& info) {
+   return info.prp_prinfo.pri_share_mode != SM_EMPTY
+          || info.prp_prinfo.pri_user_tag != 0
+          || info.prp_vip.vip_path[0] != '\0'
+          || info.prp_prinfo.pri_protection != 0
+          || info.prp_prinfo.pri_max_protection != 0;
+}
+
+void MemMapPrinter::pd_print_all_mappings(const MappingPrintSession& session) {
+
+  ProcSmapsPrinter printer(session);
+  ProcSmapsSummary summary;
+  outputStream* const st = session.out();
+  const pid_t pid = getpid();
+
+  printer.print_legend();
+  st->cr();
+  printer.print_header();
+
+  proc_regionwithpathinfo region_info_with_path;
+  MappingInfo mapping_info;
+  uint64_t address = 0;
+  int region_count = 0;
+  while (true) {
+    if (++region_count > MAX_REGIONS_RETURNED) {
+      st->print_cr("limit of %d regions reached (results inaccurate)", region_count);
+      break;
+    }
+    ::bzero(&region_info_with_path, sizeof(region_info_with_path));
```

Comment:
Should we use memset instead?

Reply:
A very good question!  Changing it now would invalidate the reviews, unfortunately.  Portability is not a concern here as it's for macOS only.  Since bzero only ever zeros memory, it _can_ be slightly faster: [one discussion](https://fdiv.net/2009/01/14/memset-vs-bzero-ultimate-showdown).